internal/state/rollback.go

```diff
@@ -86,23 +118,23 @@ func Rollback(bs BlockStore, ss Store, removeBlock bool, privValidatorConfig *co
 			Software: version.TMVersion,
 		},
 		// immutable fields
-		ChainID:       invalidState.ChainID,
-		InitialHeight: invalidState.InitialHeight,
+		ChainID:       latestState.ChainID,
+		InitialHeight: latestState.InitialHeight,
 
-		LastBlockHeight: rollbackBlock.Header.Height,
+		LastBlockHeight: lastBlockHeight,
 		LastBlockID:     rollbackBlock.BlockID,
 		LastBlockTime:   rollbackBlock.Header.Time,
 
-		NextValidators:              invalidState.Validators,
-		Validators:                  invalidState.LastValidators,
+		AppHash:         rolledBackAppHash,
+		LastResultsHash: rolledBackLastResultHash,
+
+		NextValidators:              latestState.Validators,
+		Validators:                  latestState.LastValidators,
 		LastValidators:              previousLastValidatorSet,
 		LastHeightValidatorsChanged: valChangeHeight,
 
 		ConsensusParams:                  previousParams,
 		LastHeightConsensusParamsChanged: paramsChangeHeight,
-
-		LastResultsHash: latestBlock.Header.LastResultsHash,
-		AppHash:         latestBlock.Header.AppHash,
```

Comment:
This is the main issue right? Before your PR we were using block N's hashes, while in this PR we use N-1's hashes

Reply:
yeah